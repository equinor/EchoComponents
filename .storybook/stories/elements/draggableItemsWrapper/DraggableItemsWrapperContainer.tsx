import React, { useState } from 'react';
import {
    DraggableItemsWrapper,
    DraggableItemsWrapperProps
} from '../../../../src/elements/draggableItemsWrapper/DraggableItemsWrapper';
import { DraggableItem } from '../../../../src/types/draggableItem';
import styles from './draggableItemsWrapper.module.css';

const DraggableItemsWrapperContainer: React.FC<DraggableItemsWrapperProps> = ({
    style
}: DraggableItemsWrapperProps) => {
    const [testElements, setTestElements] = useState<React.ReactNode[]>([
        <span key="1" className={styles.draggableItem}>
            1
        </span>,
        <span key="2" className={styles.draggableItem}>
            2
        </span>,
        <span key="3" className={styles.draggableItem}>
            3
        </span>,
        <span key="4" className={styles.draggableItem}>
            4
        </span>,
        <span key="5" className={styles.draggableItem}>
            5
        </span>,
        <span key="6" className={styles.draggableItem}>
            6
        </span>
    ]);

    return (
        <div className={styles.draggableOrderWrapper}>
            <div className={styles.list}>
                <DraggableItemsWrapper
                    style={style}
                    onChange={(newDragItems: DraggableItem[]) => {
                        setTestElements(newDragItems.map((n) => n.element));
                    }}
                >
                    {testElements}
                </DraggableItemsWrapper>
            </div>
        </div>
    );
};

export default DraggableItemsWrapperContainer;
