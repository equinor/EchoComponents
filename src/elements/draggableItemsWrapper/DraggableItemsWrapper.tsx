import React, { CSSProperties, useEffect, useState } from 'react';
import { ReactSortable, SortableEvent } from 'react-sortablejs';
import { DraggableItem } from '../../types/draggableItem';

const DraggableHandleSelector = 'globalDraggableHandle';

export interface DraggableItemsWrapperProps {
    style?: CSSProperties;
    onChange: (newDragItems: DraggableItem[], oldIndex: number | undefined, newIndex: number | undefined) => void;
    children: React.ReactNode[];
}

/**
 * Component that renders a wrapper for items that are draggable
 *
 * @param {DraggableItemsWrapperProps} {
 *     style: style element to override wrapper style
 *     onChange: method that will be called when elements have be reordered inside the wrapper
 *     children: list of elements that can be reordered
 * }
 * @return {*}  {JSX.Element}
 */
export const DraggableItemsWrapper: React.FC<DraggableItemsWrapperProps> = ({
    style,
    onChange,
    children
}: DraggableItemsWrapperProps): JSX.Element => {
    const [dragItems, setDragItems] = useState<DraggableItem[]>(
        children.map((element, id) => {
            return { id, element };
        })
    );

    useEffect(() => {
        setDragItems(
            children.map((element, id) => {
                return { id, element };
            })
        );
    }, [children]);

    return (
        <div style={style}>
            {dragItems.length > 0 && (
                <ReactSortable
                    animation={200}
                    handle={`.${DraggableHandleSelector}`}
                    list={dragItems}
                    onEnd={(evt: SortableEvent): void => {
                        onChange(dragItems, evt.oldIndex, evt.newIndex);
                    }}
                    setList={setDragItems}
                >
                    {dragItems.map((dragItem) => (
                        <div key={dragItem.id} className={DraggableHandleSelector}>
                            {dragItem.element}
                        </div>
                    ))}
                </ReactSortable>
            )}
        </div>
    );
};
