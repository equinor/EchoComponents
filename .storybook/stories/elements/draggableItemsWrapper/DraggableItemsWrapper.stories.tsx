import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import { DraggableItemsWrapperProps } from '../../../../src/elements/draggableItemsWrapper/DraggableItemsWrapper';
import DraggableItemsWrapperContainer from './DraggableItemsWrapperContainer';
export default {
    title: 'Elements/Functional/DraggableItemWrapper',

    component: DraggableItemsWrapperContainer,
    argTypes: {}
} as Meta;

const Template: Story<DraggableItemsWrapperProps> = (args) => <DraggableItemsWrapperContainer {...args} />;

export const Default = Template.bind({});

Default.args = {
    style: { width: '350px' }
};
