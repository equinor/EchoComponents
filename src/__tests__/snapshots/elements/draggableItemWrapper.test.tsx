import React from 'react';
import renderer from 'react-test-renderer';
import { DraggableItemsWrapper } from '../../../elements/draggableItemsWrapper/DraggableItemsWrapper';

it('renders DraggableItemsWrapper correctly', () => {
    const tree = renderer
        .create(
            <DraggableItemsWrapper onChange={jest.fn()}>
                <span key="1">1</span>
                <span key="2">2</span>
                <span key="3">3</span>
                <span key="4">4</span>
                <span key="5">5</span>
                <span key="6">6</span>
            </DraggableItemsWrapper>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
