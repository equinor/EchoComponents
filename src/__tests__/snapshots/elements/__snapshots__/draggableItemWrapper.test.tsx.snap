// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DraggableItemsWrapper correctly 1`] = `
<div>
  <div>
    <div
      className="globalDraggableHandle"
      data-id="0"
    >
      <span>
        1
      </span>
    </div>
    <div
      className="globalDraggableHandle"
      data-id="1"
    >
      <span>
        2
      </span>
    </div>
    <div
      className="globalDraggableHandle"
      data-id="2"
    >
      <span>
        3
      </span>
    </div>
    <div
      className="globalDraggableHandle"
      data-id="3"
    >
      <span>
        4
      </span>
    </div>
    <div
      className="globalDraggableHandle"
      data-id="4"
    >
      <span>
        5
      </span>
    </div>
    <div
      className="globalDraggableHandle"
      data-id="5"
    >
      <span>
        6
      </span>
    </div>
  </div>
</div>
`;
